doc/neps/nep-0022-ndarray-duck-typing-overview.rst

```diff
@@ -0,0 +1,340 @@
+realized that partial duck arrays also wanted to take advantage of
+these mixins in some cases, so even if we did have an array ABC then
+the mixins would still need some sort of separate existence. So never
+mind that argument.)
+
+Tentative duck array guidelines
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As a general rule, libraries using duck arrays should insist upon the
+minimum possible requirements, and libraries implementing duck arrays
+should provide as complete of an API as possible. This will ensure
+maximum compatibility. For example, user should prefer to rely on
+``.transpose()`` rather than ``.rollaxis()`` (which can be implemented
+in terms of transpose), but duck array authors should ideally
+implement both.
+
+If you’re trying to write a duck array, then by default this means you
+kind of need to do everything. You certainly need ``.shape``,
+``.ndim`` and ``.dtype``, but also that dtype should actually be a
+``numpy.dtype`` object, weird fancy indexing edge cases should ideally
+work, etc. Only details related to NumPy’s specific ``np.ndarray``
+implementation (e.g., ``strides``, ``data``, ``view`) are explicitly
+out of scope.
+
+A (very) rough sketch of future plans
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This proposal will help NumPy’s duck typing support, but creative use
+of array methods and existing protocols clearly does not yet suffice
+for full duck typing. We expect the need for additional protocols to
+support (at least) these features:
+
+* **Concatenating** duck arrays, which would be used internally by other
+  array combining methods like stack/vstack/hstack. The implementation
+  of concatenate will need to be negotiated among the list of array
+  arguments. We expect to use an ``__array_concatenate__`` protocol
+  like ``__array_ufunc__`` instead of multiple dispatch.
+* **Ufunc-like functions** that currently aren’t ufuncs. Many NumPy
+  functions like median, percentile, sort, where and clip could be
+  written as generalized ufuncs but currently aren’t. Either these
+  functions should be written as ufuncs, or we should consider adding
+  another generic wrapper mechanism that works similarly to ufuncs but
+  makes fewer guarantees about how the implementation is done.
+* **Random number generation** with duck arrays, e.g.,
+  ``np.random.randn()``. For example, we might want to add new APIs
+  like ``random_like()`` for generating new arrays with a matching
+  shape *and* type – though we'll need to look at some real examples
+  of how these functions are used to figure out what would be helpful.
+* **Miscellaneous other functions** such as ``np.einsum``,
+  ``np.zeros_like``, and ``np.broadcast_to`` that don’t fall into any
+  of the above categories.
+* **Checking mutability** on duck arrays, which would imply that they
+  support assignment with ``__setitem__`` and the out argument to
+  ufuncs. Many otherwise fine duck arrays* are not mutable, and it
```

Comment:
`*` was originally pointing to a footnote, so perhaps we can remove this now?

Reply:
Oh whoops, I couldn't find where the footnote was referenced, so I thought it was unused and deleted it :-).

I've re-added the original content, but rephrased using a parenthetical because I'm too lazy to figure out sphinx's footnote syntax.